core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -117,26 +117,30 @@ class BrokerMetadataListener(
       } finally {
         reader.close()
       }
-      _publisher.foreach(publish)
 
-      // If we detected a change in metadata.version, generate a local snapshot
-      val metadataVersionChanged = Option(_delta.featuresDelta()).exists { featuresDelta =>
-        featuresDelta.metadataVersionChange().isPresent
+      _bytesSinceLastSnapshot = _bytesSinceLastSnapshot + results.numBytes
+      if (_publisher.nonEmpty && shouldSnapshot()) {
```

Comment:
Perhaps we could move the publisher check into `metadataVersionChanged()`? My understanding is that `metadataVersionChanged` will always returns true until we have applied the first delta, so that would make the dependence a little clearer. Do I have that right? 

A consequence of changing that is that we could publish snapshots before the publisher has started if enough bytes have accumulated, which was the prior behavior. That would mean that we wouldn't need to update `StartPublishingEvent` below. Do you see any issues with that?

Reply:
This suggestion makes the logic clearer, but we still have to update `StartPublishingEvent` since there may be a `metadataVersionChanged`.